sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,574 @@
+﻿using System;
+using System.CodeDom.Compiler;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    /// <summary>
+    /// Generates a class that implements IFunctionMetadataProvider and the method GetFunctionsMetadataAsync() which returns a list of IFunctionMetadata. 
+    /// This source generator indexes a Function App and explicitly creates a list of DefaultFunctionMetadata (which implements IFunctionMetadata) from the functions defined
+    /// in the user's compilation. This allows the worker to index functions at build time, rather than waiting for the process to start.
+    /// </summary>
+    [Generator]
+    public class FunctionMetadataProviderGenerator : ISourceGenerator
+    {
+        public void Execute(GeneratorExecutionContext context)
+        {
+            // retreive the populated receiver 
+            if (!(context.SyntaxReceiver is SyntaxReceiver receiver))
+                return;
+
+            Compilation compilation = context.Compilation;
```

Comment:
Curious, is it expected to have this as a big class or can we break it into smaller classes?

Reply:
The [class that generates the functions.metadata file](https://github.com/Azure/azure-functions-dotnet-worker/blob/main/sdk/Sdk/FunctionMetadataGenerator.cs) for the indexing we currently use is 717 lines (which is longer than this one). I also think that source generators are usually one class (all of the components needed to create one file), but most examples I looked at are not taking on tasks this large.

Reply:
Interesting. Sounds good.